### ISSUE TITLE
expand_topic_or_service_name should use get_effective_namespace()

### ISSUE THREAD

**Issue by fmrico** [ros2/rclcpp#985]
#### Feature description

`sub_node` concept is not being correctly used in some packages that use the function `expand_topic_or_service_name()`. They use with `get_namespace()` to create publishers, instead of `get_effective_namespace()`. Probably many people do not know the existence of this method or the concept of subnode. For example, [Here](https://github.com/ros-perception/image_common/blob/42cd97418b74ca1683dcd40c7f48a711302ac3f1/image_transport/src/camera_publisher.cpp#L89). If this node is a sub_node (created with `create_sub_node()`, then the effect of the subnode does not take effect. For:

```
    color_subnode_ = create_sub_node("rgb");
    image_transport::ImageTransport color_it(color_subnode_);
    pub_color_ = color_it.advertiseCamera("image", 1);
```

The effect is:

```
$ ros2 topic list
/camera_info
/image
/image/compressed
/image/compressedDepth
/image/theora
```

Instead of adding "/rgb" in front of each topic name. 

Maybe `expand_topic_or_service_name()` should receive the rclcpp::Node, and internally, it could call `get_effective_namespace()`.

I can work on it, but I would like to know if you think it's the right solution.

#### Implementation considerations

**Comment by wjwwood** [comment]
> Here. If this node is a sub_node (created with create_sub_node(), then the effect of the subnode does not take effect.

Yeah, that looks like it is just doing the wrong thing in that case.

Unfortunately, I see the problem that these things (like image transport) take the `NodeBaseInterface` and `NodeTopicsInterface` rather than a `rclcpp::Node`, and that's a problem because the subnode mechanism exists completely within a `rclcpp::Node`. We need some way to capture this and provide it to interfaces like this when needed... I do not, for example, think the right solution is to make the image transport only take a `rclcpp::Node` object.
